apps/sms/test/marionette/lib/messages.js

```diff
@@ -256,6 +256,34 @@
           client.helper.waitForElement(SELECTORS.main);
         },
 
+        launchFromNotification: function(notification) {
+          var chromeClient = client.scope({ context: 'chrome' });
+          chromeClient.executeScript(function(origin, notification) {
+            /* global Components, Services */
+            var managerClass = Components.classes[
+              '@mozilla.org/system-message-internal;1'
+            ];
+
+            var pageURI = Services.io.newURI(
+              origin + '/index.html#notification', null, null
+            );
+
+            var manifestURI = Services.io.newURI(
+              origin + '/manifest.webapp', null, null
+            );
+
+            var systemMessageManager = managerClass.getService(
+              Components.interfaces.nsISystemMessagesInternal
+            );
+
+            systemMessageManager.sendMessage(
+              'notification', notification, pageURI, manifestURI
+            );
+          }, [ORIGIN_URL, notification]);
+
+          this.switchTo();
+        },
+
```

Comment:
> However you're currently calling internal gecko modules. I don't think it's more future-proof.

I totally agree with you that it's not future-proof, but in your proposal we'll still have ```sms-received``` system message anyway, I'll send mail to dev gaia - maybe there is "canonical" way somewhere :)

> Reminds me of my discussion with Johan :) I can tell you discuss with him a lot these days ;)

:)

> I think end to end integration tests have more value (testing user scenarios), and that more fine-grained tests can be done with unit tests.

Yeah, they have more immediate user value. Mmm not sure if unit test is the right thing here, but arguable.

> But I don't want to block on this; it's still better to have this tested than not.

Sure, I think I can try your proposal as well and see if we really like and want it :)

> However it seems like that we can forget pageURI and it should work: https://dxr.mozilla.org/mozilla-central/source/dom/messages/SystemMessageInternal.js#229-238

Hmm need to check, thanks.

> I don't think "launch" is really part of marionnette, it's part of the marionette js client library: https://github.com/mozilla-b2g/marionette-apps

Yeah, sorry I meant specifically marionette apps since it deals with web apps and manifest-based entry points look like a good for it.